packages/toast/src/ToastContainer/ToastContainer.tsx

```diff
@@ -0,0 +1,282 @@
+  const notifBarSpacing = showNotifBar
+    ? TOAST.notificationBarHeight + TOAST.gap
+    : 0;
+
+  /**
+   * Keep track of the DOM height of each toast element
+   */
+  const { toastHeights, totalStackHeight, updateToastHeights } =
+    useToastHeights({
+      stack,
+      getToastRef,
+      shouldExpand,
+    });
+
+  // Update on first render
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  useEffect(updateToastHeights, []);
+
+  /**
+   * We watch the toast container for mutations,
+   * and calculate the toast height variables once they are added to the DOM
+   */
+  useMutationObserver(
+    toastContainerRef.current,
+    {
+      childList: true,
+      attributes: true,
+      subtree: true,
+    },
+    updateToastHeights,
+    stackSize > 0,
+  );
+
+  /**
+   * Keep track of whether the toasts have transitioned in or out
+   */
+  const { isExpanded, handleTransitionExit, handleTransitionEnter } =
+    useToastTransitions({
+      containerRef: toastContainerRef,
+      getShouldExpand,
+      setHoveredState,
+      totalStackHeight,
+    });
+
+  const isInteracted = isHovered || shouldExpand;
+
+  /**
+   * When a user clicks away from the expanded stack, collapse the stack
+   */
+  useBackdropClick(
+    collapseToasts,
+    toastContainerRef,
+    isExpanded && stackSize > 0,
+  );
+
+  /**
+   * Callback passed into the InternalToast component as `onClose`
+   * Also fired when timeout timers expires
+   */
+  const handleClose = useCallback(
+    (id: ToastId, e?: CloseEvent) => {
+      const toast = getToast(id);
+      const toastRef = getToastRef(id);
+
+      if (toast && toastRef?.current) {
+        // We only self-close the toast if it's not externally controlled
+        if (!toast.isControlled) {
+          popToast(id);
+        }
+
+        toast.onClose?.(
+          // Call the close handler either with the default click event (if it exists),
+          // or a synthetic custom "timeout" event
+          e ?? createSyntheticEvent(new Event('timeout'), toastRef.current),
+        );
+      }
+    },
+    [getToast, getToastRef, popToast],
+  );
+
+  const getHandlerForId = (id: ToastId) => (e: CloseEvent) =>
+    handleClose(id, e);
+
+  /**
+   * Set & keep track of timers for each toast in the stack
+   */
+  useToastTimers({
+    stack,
+    isHovered,
+    callback: handleClose,
+  });
+
+  return (
+    <Portal className={toastPortalClassName}>
+      <div
+        ref={toastContainerRef}
+        id={regionId}
+        data-testid="lg-toast-region"
+        role="status"
+        aria-live="polite"
+        aria-relevant="all"
+        data-hovered={isHovered}
+        data-should-expand={shouldExpand}
+        data-expanded={isExpanded}
+        onMouseEnter={setHovered}
+        onMouseLeave={setUnhovered}
+        onFocus={setHovered}
+        onBlur={setUnhovered}
+        className={cx(toastContainerStyles, {
+          [toastContainerVisibleStyles]: stack.size > 0,
```

Comment:
nit: you could combine these conditionals since they both check for `stack.size > 0`
```
[cx(
  toastContainerVisibleStyles,
  getContainerStatefulStyles({
    isExpanded,
    isHovered,
    shouldExpand,
    totalStackHeight,
    topToastHeight: toastHeights[0],
    recentToastsLength: recentToasts.length,
    bottomOffset: notifBarSpacing,
  }),
)]: stack.size > 0,
```